salt/grains/smartos.py

```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+'''
+SmartOS grain provider
+
+:maintainer:    Jorge Schrauwen <sjorge@blackdot.be>
+:maturity:      new
+:depends:       salt.utils, salt.ext.six, salt.module.cmdmod
+:platform:      SmartOS
+
+.. versionadded:: nitrogen
+
+'''
+from __future__ import absolute_import
+
+# Import python libs
+import os
+import re
+import json
+import logging
+
+# Import salt libs
+import salt.utils.dictupdate
+import salt.utils.path
+import salt.utils.platform
+from salt.ext.six.moves import zip
+
+# Solve the Chicken and egg problem where grains need to run before any
+# of the modules are loaded and are generally available for any usage.
+import salt.modules.cmdmod
+
+__virtualname__ = 'smartos'
+__salt__ = {
+    'cmd.run': salt.modules.cmdmod.run,
+}
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only load when we are on SmartOS
+    '''
+    if salt.utils.platform.is_smartos():
+        return __virtualname__
+    return False
+
+
+def _smartos_computenode_data():
+    '''
+    Return useful information from a SmartOS compute node
+    '''
+    # Provides:
+    #   vms_total
+    #   vms_running
+    #   vms_stopped
+    #   vms_type
+    #   sdc_version
+    #   vm_capable
+    #   vm_hw_virt
+
+    grains = {}
+
+    # collect vm data
+    vms = {}
+    for vm in __salt__['cmd.run']('vmadm list -p -o uuid,alias,state,type').split("\n"):
+        vm = dict(list(zip(['uuid', 'alias', 'state', 'type'], vm.split(':'))))
+        vms[vm['uuid']] = vm
+        del vms[vm['uuid']]['uuid']
+
+    # set vm grains
+    grains['computenode_vms_total'] = len(vms)
+    grains['computenode_vms_running'] = 0
+    grains['computenode_vms_stopped'] = 0
+    grains['computenode_vms_type'] = {'KVM': 0, 'LX': 0, 'OS': 0}
+    for vm in vms:
+        if vms[vm]['state'].lower() == 'running':
+            grains['computenode_vms_running'] += 1
+        elif vms[vm]['state'].lower() == 'stopped':
+            grains['computenode_vms_stopped'] += 1
+
+        if vms[vm]['type'] not in grains['computenode_vms_type']:
+            # NOTE: be prepared for when bhyve gets its own type
+            grains['computenode_vms_type'][vms[vm]['type']] = 0
+        grains['computenode_vms_type'][vms[vm]['type']] += 1
+
+    # sysinfo derived grains
+    sysinfo = json.loads(__salt__['cmd.run']('sysinfo'))
+    grains['computenode_sdc_version'] = sysinfo['SDC Version']
+    grains['computenode_vm_capable'] = sysinfo['VM Capable']
+    if sysinfo['VM Capable']:
+        grains['computenode_vm_hw_virt'] = sysinfo['CPU Virtualization']
+
+    # sysinfo derived smbios grains
+    grains['manufacturer'] = sysinfo['Manufacturer']
+    grains['productname'] = sysinfo['Product']
+    grains['uuid'] = sysinfo['UUID']
+
+    return grains
+
+
+def _smartos_zone_data():
```

Comment:
Can we split this out into separate functions? The complexity of this function is quite high.

Reply:
How about splitting this in 3?
- _smartos_zone_pkgsrc_data
- _smartos_zone_pkgin_data
- _smartos_zone_image_data

?

Reply:
Perfect!

Reply:
@cachedout done